### DISCUSSION TITLE
Broad roadmap

### DISCUSSION THREAD

**Discussion by piever** [piever/SplitApplyPlot.jl#16] [General]
This is a broad roadmap to establish how to avoid duplication of code (and work) and get a reasonably general framework for "data-based" plotting in Makie. This package (SplitApplyPlot) tries to set the foundations.

## Relevant low-level types

1. `Entry`. It is a single "plottable" object, corresponding to a single trace. It encodes plot type, a mapping (data to plot attributes), as well as data independent settings.
2. `Entries`. It includes a list of `Entry`s, as well as `scales` and `labels`. `scales` and `labels` have to be consistent, so they are only encoded once. They are automatically broadened as needed as more `Entry`s are added to the `Entry` object. Various `Entries` object can be merged in a unique one.
3. `AxisEntries`. It includes a list of `Entry`s, `scales`, `labels`, and an axis on which to plot. After all `Entry`s have been collected in one unique `Entries` object, this object can be "triaged" into a grid of `AxisEntries`, using the layout information.

For an example on how to use an `AxisEntries` object constructed manually, see https://piever.github.io/SplitApplyPlot.jl/dev/generated/entries/#The-AxisEntries-type.

## From data to `Entries`

The other key ingredient is a way to pass from data to `Entries`. So far only a simple way is implemented (`splitapplyplot!`) that takes the form e.g.

```julia
ag = splitapplyplot!(
    Scatter,
    fig,
    mpg,
    :Displ => automatic => "Displacement",
    :Cty => log10 => "City miles",
    color=:Cyl => categoricalscale => "Cylinders", # attribute = variable => scale => label
)
```

The scale can (not yet fully implemented...) be used both to set a continuous transform or, in the discrete case, the order in which to list the unique values as well as the "palette" for the attribute. It is also used to determine whether a variable is discrete or continuous.

The function `splitapplyplot!` creates the `Entries` object, converts to grid of `AxisEntries` and plots it on the figure. The grid is returned and has all the information (axes objects, data, fitted scales, labels) for further tweaking.  This is demoed [here](https://piever.github.io/SplitApplyPlot.jl/dev/generated/entries/#Generating-AxisEntries-objects), where the layout is tweaked "after the fact".

## The algebra

The machinery described above already goes a long way. In particular, just by allowing the use of analyses in `splitapplyplot!` (functions that return one or more `Entry`) it should already be sufficient for most purposes. However, analyses do create some minor doubts, because they could change the number of "positional arguments" (for example if one is doing dimensionality reduction), and then it would be less clear how to pass scales along, so maybe one should pass scales separately, even though the "triple pair" syntax is quite nice. This part requires a bit more thought.

Optionally, one could build an AlgebraOfGraphics on top of it, where some "plot descriptors" (analysis, mappings, etc.) are combined formally. Then, when `draw`ing, each addend is transformed into an `Entries` objects. All `Entries` are then merged and plotted as a grid of `AxisEntries`. This should decouple syntax (formally manipulate symbolic objects with `*` and `+`), data processing (turning each addend into `Entries`), and plotting (plotting the grid of `AxisEntries`).

This decoupling is particularly important IMO as it makes it much easier to have contributors, as one can work on one of those bits without understanding the others.

## TODO list

1. Automated legend. The `scales` and `labels` object have all the required information, but as of now there is no "legend recipe". Could be ported either form AlgebraOfGraphics or TabularMakie (esp. the colorbar part should come from there).
2. Finalize discrete scale implementation (should be mostly OK already)
3. Figure out how to incorporate analyses exactly.
4. Port analyses and algebraic machinery from AlgebraOfGraphics.
5. Add all those little extra grammar of graphics things that are missing (`dodge`, `stack`, `facet_wrap`, facet labels)

The fourth point requires a little extra thought. I think that there are three reasonable ways forward:

1. Just add analyses and algebraic machinery here and slowly phase out AlgebraOfGraphics
2. Make a _very_ breaking release in AlgebraOfGraphics and have it use SplitApplyPlot as a backend.

In turn, with 1., one could decide whether to actually keep the name AlgebraOfGraphics and just replace the content with this new implementation. With 2., one should choose where some things would go. For example, do basic analyses (linear regression, density estimation, etc) belong to SplitApplyPlot or to AlgebraOfGraphics?

CC: @jkrumbiegel @greimel as we were discussing this on slack.

**Comment by jkrumbiegel** [comment]
This looks quite cool, although I can't say I fully understand it, yet. I like that you've put more emphasis on customization, as in my experience, one often needs to move away from completely "symmetric" plots, where each facet has exactly the same treatment applied. For example, I often remove certain axis labels, or add new ones specifically in some places, etc. So it's nice to have the axes more readily available.

Maybe it would be helpful to write down more mock-examples, this way we could see which cases are handled well or not so well.

About the analyses, one thing I remember from ggplot was that you have `geom`s and `stat`s, and I was confused when you should use what. If I remember correctly, as it turned out, every `geom` has an implicit `stat` and every `stat` has an implicit `geom`. So maybe the solution here is to think of the base case not as having no analysis, but just `identity`, and treating it like there's always geoms and stats together. So instead of `Scatter`, one could maybe pass `(mean, Scatter)` or something like that.

So here are some cases that we should cover, from the top of my head:

- a scatter plot, :x vs :y
- a line plot, :x vs :y
- a scatter and line plot on top of each other, :x vs :y
- a scatter plot, :x vs :y, with markersize and color controlled by two other variables
- a facet plot with one scatter in each axis, i rows and j columns
- a facet plot with one scatter in each axis, n cells wrapped into j columns
- a facet plot with one scatter and one line in each axis
- a facet plot with one scatter in each axis, and one line in only one axis (how to specify the cell)
- a scatter plot with a line fit through it
- linking all axes of a facet plot, or only link per column / row
- a normal density plot
- a vertical density plot (axis labels have to follow this, in general "flipped" plots are tricky)
- a categorical plot with tick labels set by data
- a categorical plot where tick labels from the data are overridden with some lookup function
- overridden axis labels / legend titles / legend labels
- using a log10 / log / etc scale
- a text annotation plot (difficult because the first argument is the text, and the position is given as an attribute, the system must be flexible enough to route this)
- a custom plot type wired correctly into this system, even with tricky argument / attribute orders
- an extensible way to route any data column to any plot attribute using any arbitary transformation / scaling
- an extensible way to display any plot type's special attribute in a legend
- plotting a whole facet plot into an existing figure at some arbitrary FigurePosition
- set a default plot type for a custom analysis
- change the plot type for a predefined analysis from its default (poly or line for a density plot, for example)

These are some cases that came to my mind right now, some of them are quite tricky. They don't have to be solved immediately of course, but it's good to leave avenues open to add them later, if possible.

**Reply by piever** [reply]
Thanks for the list of "things that should work", that's super helpful. A reasonable fraction of them is solvable already, but yeah, some could definitely be a bit tricky. I'll start making an "Example Gallery" doc section to collect this.

The other nice thing about the list is that it should probably help in figuring out whether the "algebra" syntax layer is worth it (depending on how much it helps in concrete examples).

**Comment by greimel** [comment]
Nice! This looks very exciting!

### Automated legends

I can take a shot at porting stuff from TabularMakie.

IIRC the implementation in AlgebraOfGraphics uses the `Legend(plts)` machinery from AbstractPlotting. That is why it cannot handle continuous styling (e.g. markersize or alpha) out of the box. Also, it is very difficult to extend.

In TabularMakie I build up the legend entries from traces, covering continuous variables and also colorbars.

It would be nice if we had something like `Legend(figpos, aog::AxisEntries; attr...)` that works like `Legend(figpos, ax)`.

### Faceting

Is `AxisEntries` meant to hold the entries for one axis, or does it hold all entries for faceted axes? If not it would be nice if it would, such that layout position is just another attribute that is stored for an axis.

That way it will be easier to implement a nice UI for legends.

### Transformations

It would be nice to allow transformations of variables and not just scales. (Which supports the DataFrames `variables => transformation => source` syntax. That would keep the door open for things like

```julia
lines(tbl, :time, [:x, :y] => ByRow(/) => "x over y")
# or
scatter(tbl, :x, :y, color = :grp => recode(...))
```

(Maybe I should open some issues?)

**Reply by piever** [reply]
> It would be nice if we had something like `Legend(figpos, aog::AxisEntries; attr...)` similar that works like `Legend(figpos, ax)`.

Yes, that's definitely the plan! A PR with the machinery you have already would be great, I think the TabularMakie set up is much more general than AlgebraOfGraphics'!

> Is AxisEntries meant to hold the entries for one axis, or does it hold all entries for faceted axes?

Both things exist (but we may wish to get rid of `AxisEntries` altogether). The `Entries` object has the list of _all traces_ with the `layout_x, layout_y` attributes in them to determine where it belongs (see e.g. `entries` [here](https://piever.github.io/SplitApplyPlot.jl/dev/generated/entries/#Transforming-and-accumulating-Entries)). When plotting it, it is transformed into a grid of `AxisEntries`, but I wonder if we could just skip that step and return separately a grid of axes and the unique `Entries` object.

I thought `AxisEntries` (with separate scales and labels per `AxisEntries`) was useful to allow for scales to vary across different `Axis`, but it may just complicate things. No harm in keeping both at the moment though.

**Reply by piever** [reply]
Re: transformations, yes, let's discuss that in a separate issue. It can be interesting especially to allow things like `layout = (:x, :y)` or `layout = :x => t -> fldmod1(t, 3)` for wrapped layout.

I still need to fully wrap up my head around scales vs transformation before or after analyses.
